### ISSUE TITLE
PHPUnit integration

### ISSUE THREAD

**Issue by mnapoli** [PHP-DI/PHP-DI#125]
Injection in PHPUnit test cases.

**Comment by moberemk** [comment]
Just something we ran in to earlier: we have a test set up for an abstract class which would normally work using the `getMockForAbstractClass` method, but that doesn't work with the DI which we need for that class. Not sure how hard it would be to accomplish but the ability to get a mocked abstract class that has had dependencies injected on it could be useful.

**Comment by mnapoli** [comment]
> the ability to get a mocked abstract class that has had dependencies injected on it could be useful.

Interesting, however I've never needed that :/ Can't you create the mock for abstract class with "not calling the constructor"? (that way you don't have to have the dependencies)

**Comment by moberemk** [comment]
Yeah this method will fail for constructor injection, but property
injection after that mock is built should run fine. Is there a setting to
not use the constructor that I'm missing in the documentation?

**Comment by mnapoli** [comment]
I was thinking of e.g.

```php
$stub = $this->getMockBuilder('SomeClass')
                     ->disableOriginalConstructor()
                     ->getMock();
```

**Comment by moberemk** [comment]
That wouldn't handle constructor injection though, since unless there's some strange PHP quirk I don't know about you can't call a constructor after instantiation without moving the actual init code out to a separate method, which basically negates the point of constructor injection IMO.

**Comment by mnapoli** [comment]
OK I think I understand better mock, you'd like a mock builder that creates a mock but that also injects dependencies? If yes then I think that's a more advanced use case honestly (you can always create the mock and provide the constructor parameters manually to PHPUnit (i.e. do the dependency injection yourself). However I can see how an automated solution could be helpful too.

**Comment by felixfbecker** [comment]
Hi, I would really like to have this feauture. I try to avoid the `@inject` annotations though because I prefer parameter injection, I think this should be supported aswell. But currently it really is a problem because you have to recreate the container in every `setUp` call, while it would make more sense to share it across tests (like a DB connection). Maybe this is even possible with `--bootstrap`, but I think a wrapper package like for the MVC frameworks would be fine.

**Comment by mnapoli** [comment]
> while it would make more sense to share it across tests (like a DB connection)

But that's the best way to have tests that affect each others, which is a nightmare to debug :(. Recreating the container in `setUp()` is a very sane approach.

**Comment by felixfbecker** [comment]
But then in every `setUp()` I would reconnect to the DB, which is totally unneeded overhead...

**Comment by mnapoli** [comment]
Right but how much time does it take? What percentage is that for each test?
